🎉 New Destination: Timeplus

jovezhong opened the pull request:
(this is to replace https://github.com/airbytehq/airbyte/pull/20391 Cannot share secrets to maintainer if the PR is sent from a github org not personal account)

## What
A new destination connector for Timeplus (works for any cloud deployment or onprem deployment of Timeplus)

close #20195 

## How
This python based destination connector will send data to Timeplus. It will create streams in Timeplus if necessary, with the exactly same schema from the source (kind of normalization, but not based on DBT). For example, if there are 4 columns in the CSV, then there will be 4 columns in the Timeplus stream.

## Recommended reading order
airbyte-integrations/connectors/destination-timeplus/destination_timeplus/destination.py

## 🚨 User Impact 🚨
no breaking change

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

To test this connector, you can create a free account at https://us.timeplus.cloud, sign in with Google/MS SSO, then create a workspace, create an API key,  then set them in the secret/UI to push data to it (tested with CSV and db)

## Tests

<details><summary><strong>Unit</strong></summary>

```
pytest -s unit_tests
=================================================== test session starts ===================================================
platform darwin -- Python 3.9.16, pytest-7.2.0, pluggy-1.0.0 -- /Users/jove/Dev/jovezhong/airbyte/airbyte-integrations/connectors/destination-timeplus/.venv/bin/python
cachedir: .pytest_cache
rootdir: /Users/jove/Dev/jovezhong/airbyte, configfile: pytest.ini
collected 1 item

unit_tests/unit_test.py::test_type_mapping PASSED

=================================================== 1 passed in 0.71s ===================================================
```

</details>

<details><summary><strong>Integration</strong></summary>

```
=================================================== test session starts ===================================================
platform darwin -- Python 3.9.16, pytest-7.2.0, pluggy-1.0.0 -- /Users/jove/Dev/jovezhong/airbyte/airbyte-integrations/connectors/destination-timeplus/.venv/bin/python
cachedir: .pytest_cache
rootdir: /Users/jove/Dev/jovezhong/airbyte, configfile: pytest.ini
collected 2 items

integration_tests/integration_test.py::test_check_valid_config {"type": "LOG", "log": {"level": "INFO", "message": "Successfully connected to https://dev.timeplus.cloud/latest"}}
PASSED
integration_tests/integration_test.py::test_check_invalid_config PASSED
```

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

jovezhong commented:
Any maintainer look into this PR? It has been quite a while

jovezhong commented:
still waiting for a reviewer

jovezhong commented:
Again, waiting for reviewers. What is the process to assign such review tickets?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=21226) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=21226) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: jovezhong<br/>:x: jerri-airbyte<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=21226) it.</sub>

marcosmarxm commented on airbyte-integrations/connectors/destination-timeplus/destination_timeplus/destination.py:
```diff
@@ -0,0 +1,122 @@
+        tp_stream.create()
+
+    @staticmethod
+    def type_mapping(v) -> str:
+        airbyte_type=v['type']
+        if type(airbyte_type) is list:
+            for t in list(airbyte_type):
+                if t!='null': 
+                    type_def={'type':t}
+                    if t=='array': type_def['items']=v['items']
+                    return DestinationTimeplus.type_mapping(type_def)
+        if airbyte_type=='number':return 'float'
+        elif airbyte_type=='integer':return 'integer'
+        elif airbyte_type=='boolean':return 'bool'
+        elif airbyte_type=='object':return 'string'
+        elif airbyte_type=='array':
+            return f"array({DestinationTimeplus.type_mapping(v['items'])})"
+        else:return 'string'
```
Can you format the code, this doesn't look it is following the PEP-8.

jovezhong replied:
I ran `Format Document` from vs code.

marcosmarxm commented on airbyte-integrations/connectors/destination-timeplus/destination_timeplus/destination.py:
```diff
@@ -0,0 +1,122 @@
+        for configured_stream in configured_catalog.streams:
+            is_overwrite = configured_stream.destination_sync_mode == DestinationSyncMode.overwrite
+            stream_exists = configured_stream.stream.name in all_streams
+            if is_overwrite and stream_exists:
+                # delete the existing stream
+                Stream(env=env).name(configured_stream.stream.name).get().delete()
+            if is_overwrite or not stream_exists:
+                # create a new stream
+                DestinationTimeplus.create_stream(env, configured_stream.stream)
```
```suggestion
            if is_overwrite: 
                if stream_exists:
                    # delete the existing stream
                    Stream(env=env).name(configured_stream.stream.name).get().delete()
                else: 
                       # create a new stream
                       DestinationTimeplus.create_stream(env, configured_stream.stream)
          else:
               raise Error("Sync mode not implemented)
```

jovezhong replied:
It's more complicated than that.  We do support is_overwrite=False

All four conditions
<img width="376" alt="image" src="https://user-images.githubusercontent.com/5076438/222022669-a070a6e8-ab64-4c6c-af40-126de69acb54.png">

My previous commits to minimize the duplicated code, but yes, it's hard to maintain. I will rewrite this part.

jovezhong replied:
pls review the new code.

jovezhong commented on airbyte-integrations/connectors/destination-timeplus/destination_timeplus/spec.json:
```diff
@@ -0,0 +1,34 @@
+    "additionalProperties": false,
+    "properties": {
+      "endpoint": {
+        "title": "Endpoint",
```
Timeplus provides fully-managed cloud as well as on-prem deployment. For public cloud, it can be https://us.timeplus.cloud and https://cloud.timeplus.com.cn. For on-prem deployments, the endpoint can be any hosted domain.

jovezhong commented on airbyte-integrations/connectors/destination-timeplus/destination_timeplus/destination.py:
```diff
@@ -0,0 +1,122 @@
+            else:
+                # ignore other message types for now
+                continue
+        pass
```
removed the pass

jovezhong commented:
Hi @marcosmarxm , thanks for the reviews. I am not quite sure about the process. I resolved the review comments. Feel free to reopen them if they need more changes. I added 3 commits to address your review comments one by one. Hopefully it should make it a bit easier to check the code change.

jovezhong commented:
Hi @marcosmarxm, could you please take the other look? Thanks

jovezhong commented:
I don't want to ask this every week, but please please give some feedback @marcosmarxm

marcosmarxm commented:
Sorry the long delay @jovezhong I added this to our team backlog. Hope to have a slot this week.

jovezhong commented:
Really appreciate it. I work in Europe timezone this week, should be able to respond to your review comments within 12 hours

jovezhong commented:
I am attending Data Council this week at Austin. Will stop by AirByte booth to see whether I can get some help on this PR review.

jovezhong commented:
@marcosmarxm @CLAassistant , please check the PR

jovezhong commented:
Hi @marcosmarxm , thanks for looking into this. I think I have addressed all review comments. Please double check. I noticed the 'Add labels to github PRs` is still failing. Let me know what else need to be done. Thanks.

jovezhong commented:
I noticed the label changes. Anything I can help to get the PR merged?

marcosmarxm commented on airbyte-integrations/connectors/destination-timeplus/integration_tests/integration_test.py:
```diff
@@ -0,0 +1,63 @@
+    outcome = DestinationTimeplus().check(AirbyteLogger(), {"secret_key": "not_a_real_secret"})
+    assert outcome.status == Status.FAILED  
+
+def test_write(config: Mapping):
+    test_schema = {"type": "object", "properties": {"str_col": {"type": "str"}, "int_col": {"type": "integer"}}}
+
+    test_stream = ConfiguredAirbyteStream(
+        stream=AirbyteStream(name="test_stream", json_schema=test_schema, supported_sync_modes=[SyncMode.incremental]),
+        sync_mode=SyncMode.incremental,
+        destination_sync_mode=DestinationSyncMode.append,
+    )
+
+    records = [AirbyteMessage(
+        type=Type.RECORD, record=AirbyteRecordMessage(stream="test_stream", data={
+            "str_col": "example",
+            "int_col": 1,
+        }, emitted_at=0)
+    )]
+    dest = DestinationTimeplus()
+    dest.write(
+        config=config, 
+        configured_catalog=test_stream,
+        input_messages=records
+    )
```
Sorry the delay @jovezhong I added this function to test the write function. I ran it and it's not sending data to Timeplus. Please make sure it works.

marcosmarxm replied:
Also I tried the connector locally and it's not working.
![image](https://github.com/airbytehq/airbyte/assets/5154322/e67898ba-0e12-42e4-8d6a-f5c70c9729ed)

jovezhong replied:
I will be offline for 1 hour, then join a meeting.

Take a quick look at the error message. It is trying to add an event in Timeplus for `campaigns` stream for `lastProcessingFinishedAt` column, but this column is not defined. There is only a `lastUpdatedTime` column. Not sure whether hubspot will produce event with extra column which is not defined in the schema.

I will take a look later today after other meetings. In the meanwhile, you may try other sources such as CSV.

I have a daily job to push Hubspot to Timeplus via this airbyte connector, but mainly for contacts/companies, not campaigns

marcosmarxm replied:
Anything please ping me on Slack.

jovezhong commented:
>  FileNotFoundError: [Errno 2] No such file or directory: 'secrets/config.json

Need to set the credential in the file, such as
```json
{
    "endpoint": "https://us.timeplus.cloud/d22355",
    "apikey": "aP...3A3ebrFn"
}
```

marcosmarxm commented:
/test connector=connectors/destination-timeplus

marcosmarxm commented:
@jovezhong can you fix:
❌ - check_documentation_file_exists
❌ - check_changelog_entry_is_updated
❌ - check_connector_icon_is_available

jovezhong commented:
Sure, @marcosmarxm , happy to add those files. I checked the log and cannot find the exact file they are looking for? Can you share any example or docs about those 3 checks? Thanks.

❌ - check_documentation_file_exists
❌ - check_changelog_entry_is_updated
❌ - check_connector_icon_is_available

jovezhong commented:
Thank you @marcosmarxm for the review and merge!
